File: java/graph/shortest_path/Dijkstra.java
package graph.shortest_path;

import java.util.Arrays;
import java.util.Comparator;
import java.util.PriorityQueue;
import java.util.Queue;

public class Dijkstra {
    public int[] search(VertexNode[] graphs, int start) {
        boolean[] visited = new boolean[graphs.length];
        Arrays.fill(visited, false);
        Queue<HeapNode> queue = new PriorityQueue<>(cmp);
        queue.offer(new HeapNode(start, 0));

        int[] result = new int[graphs.length];
        Arrays.fill(result, Integer.MAX_VALUE);
        while (!queue.isEmpty()) {
            HeapNode heap = queue.poll();
            int index = heap.index;
            int weight = heap.weight;
            if (visited[index]) {
                continue;
            }

            result[index] = weight;
            visited[index] = true;

            EdgeNode edge = graphs[index].next;
            while (null != edge) {
                int edgeIndex = edge.index;
                int edgeWeight = weight + edge.weight;
                if (!visited[edgeIndex]) {
                    queue.offer(new HeapNode(edgeIndex, edgeWeight));
                }
                edge = edge.next;
            }
        }
        return result;
    }

    private class HeapNode {
        public int index = 0;
        public int weight = 0;

        public HeapNode(int index, int weight) {
            this.index = index;
            this.weight = weight;
        }
    }

    private static Comparator<HeapNode> cmp = new Comparator<HeapNode>() {
        @Override
        public int compare(HeapNode node1, HeapNode node2) {
            return node1.weight - node2.weight;
        }
    };

    public class EdgeNode {
        public int index = 0;
        public int weight = 0;
        public EdgeNode next = null;

        public EdgeNode(int index, int weight) {
            this.index = index;
            this.weight = weight;
        }
    }

    public class VertexNode {
        public int value = 0;
        public EdgeNode next = null;

        public VertexNode(int value) {
            this.value = value;
        }
    }
}

File: py/graph/shortest_path/Dijkstra.py
#!/usr/bin/env python
# -*- coding:gbk -*-

import heapq, sys
from typing import List

class EdgeNode:
    def __init__(self, index: int, weight: int):
        self.index, self.weight, self.next = index, weight, None

class VertexNode:
    def __init__(self, value: int):
        self.value, self.next = value, None

class Dijkstra:
    def search(self, graphs: List[VertexNode], start: int) -> List[int]:
        visited, queue = [False for _ in range(len(graphs))], [self.__HeapNode(start, 0)]
        result = [sys.maxsize for _ in range(len(graphs))]
        while queue:
            heap = heapq.heappop(queue)
            index, weight = heap.index, heap.weight
            if visited[index]:
                continue

            result[index], visited[index] = weight, True
            edge = graphs[index].next
            while edge:
                edgeIndex, edgeWeight = edge.index, weight + edge.weight
                if not visited[edgeIndex]:
                    heapq.heappush(queue, self.__HeapNode(edgeIndex, edgeWeight))
                edge = edge.next
        return result

    class __HeapNode:
        def __init__(self, index: int, weight: int):
            self.index, self.weight = index, weight

        def __lt__(self, other):
            return self.weight < other.weight
